### PR TITLE
Properly track (and mark) truncated store entries

### PR DESCRIPTION
**Pull request by eduard-bagdasaryan**
## Responses with truncated bodies

Squid used an error-prone approach to identifying truncated responses:
The response is treated as whole[1] unless somebody remembers to mark it
as truncated. This dangerous default naturally resulted in bugs where
truncated responses are treated as complete under various conditions,

This change reverses that approach: Responses not explicitly marked as
whole are treated as truncated. This change affects all Squid-server
FwsState-dispatched communications: HTTP, FTP, Gopher, and WHOIS. It
also affects responses received from the adaptation services.

Squid still tries to deliver responses with truncated bodies to clients
in most cases (no changes are expected/intended in that area).

[1]: A better word to describe a "whole" response would be "complete",
    but several key Squid APIs use "complete" to mean "no more content
    is coming", misleading developers into thinking that a "completed"
    response has all the expected bytes and may be cached/shared/etc.

## Responses with truncated HTTP headers

This change also starts treating responses with truncated HTTP headers
(i.e. no CRLF after all the field-lines) as malformed, replacing them
with an HTTP 502 ERR_INVALID_RESP error (same as any other HTTP response
with malformed headers would get).

Since Bug 2879 (commit da6c841 and earlier v2-only changes), Squid was
"fixing" such truncated headers by adding an extra CRLF sequence and
re-parsing them. Depending on the truncation circumstances, that old
workaround could result in rather bad outcomes for Squid and its
clients. Hopefully, we no longer need to work around such bugs. If we
do, we should do it in a safer manner and with admin permission.

## Related access-logging improvements

Transactions that failed due to origin server or peer timeout (a common
source of truncation) are now logged with a _TIMEOUT %Ss suffix and
ERR_READ_TIMEOUT/WITH_SRV %err_code/%err_detail.

Transactions prematurely canceled by Squid during client-Squid
communication (usually due to various timeouts) now have WITH_CLT
default %err_detail. This detail helps distinguish otherwise
similarly-logged problems that may happen when talking to the client or
to the origin server/peer.

## Other

FwdState now (indirectly) complete()s truncated entries _after_
releasing/adjusting them so that invokeHandlers() and others do not get
a dangerous glimpse at a seemingly OK entry before its release().

### CODE REVIEW

File: src/FwdState.cc [hunk 1]
```diff
@@ -1173,9 +1217,12 @@ FwdState::dispatch()
 
 #if USE_OPENSSL
     if (request->flags.sslPeek) {
+        // we were just asked to peek at the server, and we did that
         CallJobHere1(17, 4, request->clientConnectionManager, ConnStateData,
                      ConnStateData::httpsPeeked, ConnStateData::PinnedIdleContext(serverConnection(), request));
         unregister(serverConn); // async call owns it now
+        flags.dont_retry = true; // we gave up forwarding control
+        flags.abandonEntry = true; // TODO: or should we abort it instead?
```
Comment (eduard-bagdasaryan) [inline src/FwdState.cc:1225]: The overlapping names of the existing StoreEntry::abandon()  and the new 'abandonEntry' flag may be confusing and raise (maybe stupid) questions like "why can't we immediately StoreEntry::abandon() instead of remembering this state with a specific flag"?

I would rather agree with this TODO - can we simply call entry->abort() here and in dispatch() instead of juggling with these flags? In this case, we would not need to add flags.abandonEntry, I guess.

Comment (rousskov) [reply src/FwdState.cc:1225]: The overlap is somewhat intentional and, yes, it kind of indicates a problem with this code. I do not yet know whether we should use this small hack or fix this "properly". The answer depends on what a proper fix would require.

> can we simply call entry->abort() here and in dispatch() instead of juggling with these flags?

The answer to that (and TODO) question requires research. I do not know what happens when these SslBump-related entries are abort()ed, especially with respect to invokeHandlers() effects. See commit 1eb5d56 for a more detailed list of questions/concerns to research.

Comment (eduard-bagdasaryan) [reply src/FwdState.cc:1225]: We should have an empty entry here with the ENTRY_FWD_HDR_WAIT set. If so, invokeHandlers() will do nothing.

File: src/gopher.cc [hunk 2]
```diff
@@ -807,6 +813,10 @@ gopherReadReply(const Comm::ConnectionPointer &conn, char *buf, size_t len, Comm
 
         entry->timestampsSet();
         entry->flush();
+
+        if (!gopherState->len && !gopherState->overflowed)
```
Comment (eduard-bagdasaryan) [inline src/gopher.cc:817]: It is unclear what the first part of this condition means. A reader may be confused, assuming that gopherState->len means a kind of 'some bytes' received, which contradicts the statement below. Probably we should document gopherState->len (at least). It is used in gopherToHTML() - but I could not suggest a suitable definition based on its current usage.

File: src/FwdState.cc (hunk 1 above)
Comment (rousskov) [reply src/FwdState.cc:1225]: This needs to be carefully researched/tested/investigated. We should not dismiss these concerns based on superficial analysis.

```C++
StoreEntry::abort()
{
...
    // allow the Store clients to be told about the problem
    EBIT_CLR(flags, ENTRY_FWD_HDR_WAIT);
...
    invokeHandlers();
}
```

File: src/gopher.cc (hunk 2 above)
Comment (rousskov) [reply src/gopher.cc:817]: That looks like a branch bug to me, actually: Should not this be a positive `gopherState->len` test? 

```suggestion
        if (gopherState->len && !gopherState->overflowed)
```

If yes, that condition would match the "some bytes" in the corresponding markStoredReplyAsWhole() reason nicely. We could add something like `without overflowing` to that reason, but I think we do not have to document the _lack_ of problems in common/regular cases (or it may get out of hand because there are a lot of things that usually do _not_ go wrong.

If you agree that this is a bug, please reverse the condition. It would be a nice additional proof that these reason strings are useful!

File: src/FwdState.cc (hunk 1 above)
Comment (eduard-bagdasaryan) [reply src/FwdState.cc:1225]: To investigate whether/how StoreEntry::invokeHandlers() is called (in the 'splice during step3' test) and what effects it produces, I ran the test with 3 Squids:
1. this PR:
No handlers are invoked because the list of clients is cleared in storeUnregster(). Here is the call stack:
```
StoreEntry::invokeHandlers() at store_client.cc:758 0x555555b1955c	
StoreEntry::abort() at store.cc:1,149 0x555555b0e480	
storeUnregister() at store_client.cc:742 0x555555b194ae	
Ssl::ServerBump::~ServerBump() at ServerBump.cc:54 0x555555cf7438	
ConnStateData::~ConnStateData() at client_side.cc:680 0x555555a482ac	
Http::One::Server::~Server() at Http1Server.h:26 0x555555c0a2fc	
Http::One::Server::~Server() at Http1Server.h:26 0x555555c0a368	
AsyncJob::callEnd() at AsyncJob.cc:146 0x555555c2dfe7	
```
2. master:
No handlers were invoked in my simple 1-client simple test because sc->_callback.pending() returned false.
```
StoreEntry::invokeHandlers() at store_client.cc:774 0x555555b21156	
StoreEntry::complete() at store.cc:1,071 0x555555b159e8	
FwdState::complete() at FwdState.cc:561 0x555555998935	
FwdState::connectedToPeer() at FwdState.cc:1,013 0x55555599c0f0	
```
3. patched PR:
I commented the `flags.abandonEntry = true;` in connectedToPeer() and added `entry->abort()`. The result was the same as in (2), with the difference that invokeHandlers() was initiated by abort().

So it looks like we use abort() in connectedToPeer() without problems. Note that I did not test another similar path, in  FwdState::dispatch().

Comment (rousskov) [reply src/FwdState.cc:1225]: > 2. master: No handlers were invoked in my simple 1-client simple test because sc->_callback.pending() returned false.

> 3. patched PR: same as in (2), with the difference that invokeHandlers() was initiated by abort().

Hm... Please check why there is no callback in those cases. I would expect the Ssl::ServerBump object to be a registered/active StoreEntry client that expects to be notified about StoreEntry changes.

Comment (eduard-bagdasaryan) [reply src/FwdState.cc:1225]: > Ssl::ServerBump object to be a registered/active StoreEntry client 

The client is stored in ServerBump::sc, to be more precise. This object is indeed created, but its _callback field remains nil. It acquires its value in store_client::copy(), but we do not get there in this test.

File: src/gopher.cc (hunk 2 above)
Comment (eduard-bagdasaryan) [reply src/gopher.cc:817]: As we discussed, the '!gopherState->len' should work, i.e., we don't need to fix this line:
1. if we are on the gopherToHTML() path, non-zero len means an unterminated line received from the server (Gopher protocol requires all lines end with CRLF)
2. Otherwise, len remains zero (we do not look for CRLF).

However, I have some questions about it. For example, why do we consider a response 'complete' in (2) having the last line without a terminating CRLF? Also, the protocol requires '.' at the end of the message (i.e., messages are incomplete without it), but I read that many Gopher servers do not follow this rule.

Comment (rousskov) [reply src/gopher.cc:817]: I think we are simply relying on the existing gopher response parsing code.

If that existing code considers the line complete, then we are not going to judge/improve it in this PR. I assume that `gopherState->len` is zero when that code considers the last line complete, but please correct me if I am wrong.

As for the trailing `.`, I think we should again follow the existing code lead: If that existing code is waiting for that period, then we ought to reflect that in our new condition. Otherwise, we are not going to overwrite that code decision. In other words, we simply leave that decision to that code, broken or not.

File: src/FwdState.cc (hunk 1 above)
Comment (rousskov) [reply src/FwdState.cc:1225]: > ServerBump::sc ... is indeed created, but its _callback field remains nil. It acquires its value in store_client::copy()

That sounds like a Squid bug to me: The registered StoreEntry client probably wants to be notified about abort()s even if they happen before the first copy(). The client is waiting for stored data and should stop waiting if no data is coming. We may leave the arguably out-of-scope bug as is, but I do not want our code to _rely_ on it.

Please see if you can temporary hack the code so that abort() creates _callback (if needed) and calls store_client::callback() with an error, as expected. Feel free to add/call StoreEntry::invokeHandlers2() for that purpose. What happens then?

Comment (eduard-bagdasaryan) [reply src/FwdState.cc:1225]: >  temporary hack the code so that abort() creates _callback 

I don't think we need this: after examining ServerBump::sc I noticed that it is not passed to storeClientCopy() and other methods. It is documented as a " dummy client to prevent entry trimming", and it's callback is never assigned.

File: src/gopher.cc (hunk 2 above)
Comment (eduard-bagdasaryan) [reply src/gopher.cc:817]: > If that existing code considers the line complete

The old code does not consider an unterminated line complete (which is still in GopherStateData::buf) after receiving EOF, i.e., it does not add this line to the entry. On the other hand, the old code (upon receiving EOF) does not check that there is still buffered data (flushing it to the entry)  and treats the already received entry as 'complete', by calling both gopherEndHTML() and fwd->complete(). However, this is the place we are fixing here: we do not want to pretend that the entry is complete if it is not. Another approach would be to append the buffered (unterminated) line to the entry and declare the entry as 'complete' - but this would, even more, violate the Gopher specification (it is already violated by not looking for the terminating '.').

File: src/FwdState.cc (hunk 1 above)
Comment (rousskov) [reply src/FwdState.cc:1225]: > It is documented as a "dummy client to prevent entry trimming".

OK, that is useful to know, but I remain concerned that changes (perhaps even bug fixes) in the ServerBump code will enable [StoreEntry::abort()] notifications. Perhaps we can attack this from a different angle: Does that dummy client or surrounding ServerBump code has any methods to receive Store notifications? In other words, if we start passing `sc` to storeClientCopy() and friends, what function will invokeHandlers() call?

If there is no such function at all right now, then we can address my concerns by saying that it would be up to those who add that function to implement it correctly.

File: src/gopher.cc (hunk 2 above)
Comment (rousskov) [reply src/gopher.cc:817]: > Alex: If that existing code considers the line complete

> Eduard: The old code does not consider an unterminated line complete

Yes, and that is fine:

* Existing parsing code considers the line complete if `gopherState->len` is zero. Our changes match that part of the algorithm.
* Existing parsing code considers the line incomplete if `gopherState->len` is positive. Our changes match that part of the algorithm.
* Existing entry completion code declares the entry complete regardless of `gopherState->len`. That is the bug we are fixing.

In summary, nothing to change here besides adding documentation for `len`, right?

File: src/FwdState.cc (hunk 1 above)
Comment (eduard-bagdasaryan) [reply src/FwdState.cc:1225]: > if we start passing sc to storeClientCopy() and friends, what function will invokeHandlers() call

Eventually it calls store_client::Callback::callback_handler, with void `*data, StoreIOBuffer result` parameters. There are several such callbacks, for example, clientReplyContext::SendMoreData(...), but none of them 'belongs' to the ServerBump's surroundings.

File: src/gopher.cc (hunk 2 above)
Comment (eduard-bagdasaryan) [reply src/gopher.cc:817]: > nothing to change here besides adding documentation for len, right?

Agreed.

File: src/FwdState.cc (hunk 1 above)
Comment (rousskov) [reply src/FwdState.cc:1225]: I agree. Based on your research, I think we should address that TODO in this branch: Please replace flags.abandonEntry with StoreEntry::abort() calls and retest.

File: src/gopher.cc (hunk 2 above)
Comment (rousskov) [reply src/gopher.cc:820]: I added that documentation at 867b855.

Comment (rousskov) [reply src/gopher.cc:820]: BTW, `after receiving/storing some bytes` refers to the `!entry->isEmpty()` condition implied by the previous `if` statement. I think it is more-or-less decipherable from the surrounding code, but we can also add an "EOF after we stored (and, hence, received/parsed) some bytes" comment right after our `if` statement if you think it is needed.

File: src/FwdState.cc (hunk 1 above)
Comment (eduard-bagdasaryan) [reply src/FwdState.cc]: Done(7948c20) and re-tested with Daft.

**Comment by rousskov** [comment]
This became official PR 909.
